nessie/src/main/java/org/apache/iceberg/nessie/NessieCatalog.java

```diff
@@ -0,0 +1,431 @@
+import java.util.Set;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.apache.hadoop.conf.Configurable;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.iceberg.BaseMetastoreCatalog;
+import org.apache.iceberg.TableMetadata;
+import org.apache.iceberg.TableOperations;
+import org.apache.iceberg.catalog.Namespace;
+import org.apache.iceberg.catalog.SupportsNamespaces;
+import org.apache.iceberg.catalog.TableIdentifier;
+import org.apache.iceberg.exceptions.AlreadyExistsException;
+import org.apache.iceberg.exceptions.CommitFailedException;
+import org.apache.iceberg.exceptions.NamespaceNotEmptyException;
+import org.apache.iceberg.exceptions.NoSuchNamespaceException;
+import org.apache.iceberg.exceptions.NoSuchTableException;
+import org.apache.iceberg.relocated.com.google.common.base.Joiner;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
+
+/**
+ * Nessie implementation of Iceberg Catalog.
+ *
+ * <p>
+ *   A note on namespaces: Nessie namespaces are implicit and do not need to be explicitly created or deleted.
+ *   The create and delete namespace methods are no-ops for the NessieCatalog. One can still list namespaces that have
+ *   objects stored in them to assist with namespace-centric catalog exploration.
+ * </p>
+ */
+public class NessieCatalog extends BaseMetastoreCatalog implements AutoCloseable, SupportsNamespaces, Configurable {
+
+  private static final Joiner SLASH = Joiner.on("/");
+  public static final String NESSIE_WAREHOUSE_DIR = "nessie.warehouse.dir";
+  private NessieClient client;
+  private String warehouseLocation;
+  private Configuration config;
+  private UpdateableReference reference;
+  private String name;
+
+  /**
+   * Try to avoid passing parameters via hadoop config. Dynamic catalog expects Map instead
+   *
+   * todo replace with #1640 style constructor
+   */
+  public NessieCatalog() {
+  }
+
+  /**
+   * Create a catalog with a known name from a hadoop configuration.
+   */
+  public NessieCatalog(String name, Configuration config, String ref, String url, String warehouseLocation) {
+    this.config = config;
+    this.name = name == null ? "nessie" : name;
+    init(ref, url, warehouseLocation);
+  }
+
+  private void init(String ref, String url, String inputWarehouseLocation) {
+    this.client = NessieClient.withConfig(s -> {
+      if (s.equals(NessieClient.CONF_NESSIE_URL)) {
+        return url == null ? config.get(s) : url;
+      }
+      return config.get(s);
+    });
+
+    this.warehouseLocation = inputWarehouseLocation == null ? getWarehouseLocation(config) : inputWarehouseLocation;
+
+    final String requestedRef = ref != null ? ref : config.get(NessieClient.CONF_NESSIE_REF);
+    this.reference = get(requestedRef);
+  }
+
+  private static String getWarehouseLocation(Configuration config) {
+    String nessieWarehouseDir = config.get(NESSIE_WAREHOUSE_DIR);
```

Comment:
I don't think that any configuration should come from the Hadoop `Configuration` unless it is used for a Hadoop component, like `HadoopFileIO`. Can you initialize this from the catalog config passed to `initialize`?

Reply:
Hey @rdblue these parameters are initialised in the `initialize` method (I have moved the method up to near the constructor as it was hidden at teh bottom of the class). The initialisation uses the passed `options` where possible and falls back to `Configuration` if not found. This is to make it compatible with Spark2/3 `IcebergSource`. However I am happy to remove this once `IcebergSource` supports custom catalogs (which I hope to tackle next if its not already being worked on)

Reply:
We support the Hive warehouse property because that's how to set it up in Hive. Since we are introducing new configuration for Nessie, I'd really rather not make it so people can depend on using the Hadoop config. Then we will never be able to get rid of it.

Reply:
Ok, removed. The consequence of this is the custom catalog work for `IcebergSource` has to be done before the next release if we want a valid/usable nessie in the release